internal/authentication/ldap_client_factory.go

```diff
@@ -117,162 +121,393 @@ type PooledLDAPClientFactory struct {
+			return client, nil
 		}
 	}
+	return nil, fmt.Errorf("timeout waiting for client from LDAP pool")
+}
+
+// must be called EXCLUSIVELY by Initialize() or poolManager()
+func (f *PooledLDAPClientFactory) tryAddPooledClient(pC chan *LDAPClientPooled ) (bool) {
+	attempts := f.config.Pooling.Retries
+	sleep := f.config.Timeout / 8 // initial sleep duration, trade-off between responsiveness and overload
+
+	for !f.isClosing() {
+		start := time.Now()
+		client, err := f.new() // NOTE: new() adheres to config.Timeout, not config.Pooling.Timeout!
+		if err == nil {
+			// Shorten client lifetime by up to 10% to avoid mass exodus events
+			fuzzFactor := 0.90 + 0.1*rand.Float64() // [0.90..1.00)
+			client.expiresAt = time.Now().Add(time.Duration(float64(f.clientLifetime) * fuzzFactor))
+
+			select {
+			case pC <- client:
+				active := f.activeCount.Add(1)
+				logging.Logger().Debugf("Added new client to LDAP pool. Active: %d, Available: %d", active, len(pC))
+				return true
+			default: // shouldn't happen
+				logging.Logger().Warn("LDAP pool closed or full, closing newly created client")
+				_ = client.Close()
+				return false
+			}
+		}
+		logging.Logger().Debugf("Failed to create new client for LDAP pool. Attempt %d/%d, Elapsed: %s: %v",
+			f.config.Pooling.Retries - attempts + 1, f.config.Pooling.Retries + 1, time.Since(start), err)
+
+		if attempts == 0 {
+			logging.Logger().Warningf("Exceeded maximum attempts (%d) for creating pooled LDAP client; last error: %v",
+				f.config.Pooling.Retries + 1, err)
+			return false
+		}
 
-	ctx, cancel := context.WithTimeout(ctx, f.config.Pooling.Timeout)
-	defer cancel()
+		logging.Logger().Tracef("Sleeping %s before next attempt to create pooled LDAP client", sleep)
+		time.Sleep(sleep)
+		sleep *= 2
+		attempts--
+	}
+	return false // pool is closing
+}
 
+// thread-safe, does not block
+func (f *PooledLDAPClientFactory) wakeupPoolManager() {
+	wakeC := f.wakeup
+	if wakeC == nil { return } // pool is closing or not initialized
 	select {
-	case <-ctx.Done():
-		return nil, ctx.Err()
-	case client = <-f.pool:
-		if client.IsClosing() || client.Client == nil {
-			for {
-				select {
-				case <-ctx.Done():
-					return nil, ctx.Err()
-				default:
-					if client, err = f.new(); err != nil {
-						time.Sleep(f.sleep)
-
-						continue
-					}
-
-					return client, nil
+	case wakeC <- struct{}{}:
+	default: // Already signaled, do nothing
+	}
+}
+
+// goroutine that maintains the pool of LDAP clients
+func (f *PooledLDAPClientFactory) poolManager() {
+
+	pCtx, pC, wakeC, retC := f.ctx, f.pool, f.wakeup, f.closed
+	if pC == nil || pCtx == nil || wakeC == nil || retC == nil {
+		logging.Logger().Error("LDAP pool manager cannot start: some components are not initialized")
+		close(f.closed)
+		return
+	}
+
+	logging.Logger().Infof("LDAP Pool Manager started: minPoolSize=%d, maxPoolSize=%d, retries=%d, timeout=%v",
```

Comment:
Please remove call to logging.Logger() and initialize a fielded logger like in #10383.